### PR TITLE
Implementation of a generic image handler

### PR DESCRIPTION
**Pull request by jkrasting**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -53,3 +53,4 @@ install:
 script:
   - py.test om4labs
   - black --check om4labs
+  - cd testing; make all
```
Comment (raphaeldussin) [inline .travis.yml:56]: this will fail because travis won't let you copy that much data over

File: ci/requirements.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 black
 dask
-intake
-matplotlib
+intake-xarray
+matplotlib>=2.0.2,<3.0
```
Comment (raphaeldussin) [inline ci/requirements.txt:4]: can't restrict to matplotlib 2.X

File: om4labs/catalogs/OM4p125_catalog_gfdl.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    description: "File containing basin masks on native model grid"
+    driver: netcdf
+    args:
+      urlpath: '/archive/Raphael.Dussin/datasets/OM4p125/OM4p125_grid_20200213_merged_bathy_unpacked/ocean_static_d2.nc'
+      xarray_kwargs:
+        decode_times: False
+    metadata:
+      origin_url: ''
+
+  ocean_hgrid:
```
Comment (raphaeldussin) [inline om4labs/catalogs/OM4p125_catalog_gfdl.yml:16]: we're gonna need hgrid and hgrid_d2, same for basin, topo,...

File: om4labs/diags/generic_annual_bias_1x1deg/generic_annual_bias_1x1deg.py [hunk 4]
```diff
@@ -215,22 +224,18 @@ def run(dictArgs):
...
     return imgbufs
 
 
 def plot(x, y, area, model, obs, dictArgs):
     """meta plotting function"""
 
-    streamdiff = True if dictArgs["stream"] == "diff" else False
-    streamcompare = True if dictArgs["stream"] == "compare" else False
-    streamnone = True if dictArgs["stream"] is None else False
+    streamdiff = True if dictArgs["style"] == "diff" else False
```
Comment (raphaeldussin) [inline om4labs/diags/generic_annual_bias_1x1deg/generic_annual_bias_1x1deg.py:236]: I didn't like that part of my code, maybe there's a better solution

File: om4labs/diags/moc/moc.py [hunk 5]
```diff
@@ -100,15 +104,29 @@ def generate_basin_masks(basin_code, basin=None):
     return mask
 
 
-def arguments(cliargs=None):
+class DefaultDictParser(argparse.ArgumentParser):
```
Comment (raphaeldussin) [inline om4labs/diags/moc/moc.py:107]: move to om4common?

File: om4labs/diags/moc/moc.py [hunk 6]
```diff
@@ -157,17 +185,36 @@ def arguments(cliargs=None):
         "-t", "--topog", type=str, default=None, help="Path to topog file"
     )
 
-    return parser.parse_args(cliargs)
+    if template is True:
```
Comment (raphaeldussin) [inline om4labs/diags/moc/moc.py:188]: I don't understand the template purpose

File: om4labs/diags/moc/moc.py [hunk 7]
```diff
@@ -157,17 +185,36 @@ def arguments(cliargs=None):
...
-    print("Reading input data")
+    infile = dictArgs["infile"]
+
+    if dictArgs["model"] is not None:
+        # use dataset from catalog, either from command line or default
+        cat_platform = (
+            f"catalogs/{dictArgs['model']}_catalog_{dictArgs['platform']}.yml"
+        )
+        catfile = pkgr.resource_filename("om4labs", cat_platform)
+        cat = intake.open_catalog(catfile)
```
Comment (raphaeldussin) [inline om4labs/diags/moc/moc.py:205]: will fail if ressource is not available

File: om4labs/m6plot/xyplot.py [hunk 8]
```diff
@@ -72,6 +74,8 @@ def xyplot(
     # set visual backend
     if interactive is False:
         plt.switch_backend("Agg")
```
Comment (raphaeldussin) [inline om4labs/m6plot/xyplot.py:76]: this is probably machine specific

File: .travis.yml (hunk 1 above)
Comment (jkrasting) [reply .travis.yml:56]: It sort of worked ...

File: ci/requirements.txt (hunk 2 above)
Comment (jkrasting) [reply ci/requirements.txt:4]: Agree - we need to fix the vertical split scale issue.

File: om4labs/diags/moc/moc.py (hunk 5 above)
Comment (jkrasting) [reply om4labs/diags/moc/moc.py:107]: Hmm ... I thought I did.  Let me check correct and if not I will.

File: om4labs/diags/moc/moc.py (hunk 6 above)
Comment (jkrasting) [reply om4labs/diags/moc/moc.py:188]: The template functionality returns a generic dict() of options.  Dora can modify that dict() and use it to call run() directly.

File: om4labs/diags/moc/moc.py (hunk 7 above)
Comment (jkrasting) [reply om4labs/diags/moc/moc.py:205]: What's your suggestion?  Check to see if the catalog exists and fail more gracefully?

File: om4labs/m6plot/xyplot.py (hunk 8 above)
Comment (jkrasting) [reply om4labs/m6plot/xyplot.py:76]: Agg is a headless backend.  I'm pretty confident it is universal.

File: om4labs/diags/moc/moc.py (hunk 7 above)
Comment (raphaeldussin) [reply om4labs/diags/moc/moc.py:197]: exactly!

**Comment by jkrasting** [comment]
@raphaeldussin - just checking in on this PR.  Is this ready to go?

**Comment by raphaeldussin** [comment]
things to do:

- [ ] remove requirements on matplotlib 2 now that we have a fix for splitScale
- [ ]  standard_grid_cell_area looks like a duplicate of a function I already coded
- [ ] can we put the intake catalogs and testing in 2 other separate PRs?

**Comment by jkrasting** [comment]
@raphaeldussin --  I think we are good now.

- I separated out the testing and intake commits into separate PRs:  #27, #28 
- I created issue #29 to document the spherical grid area functions - let's fix that once this behemoth PR is merged in

**Pull request merged**